### DISCUSSION TITLE
dimmed code

### DISCUSSION THREAD

**Discussion by dfgordon** [microsoft/vscode-discussions#1350] [Extension Development QnA] [answered]
Some extensions apply a dimming effect to lines of code that the extension has determined are inactive, e.g. a C extension might dim something made inactive by `#ifdef`.  I would have thought this is a semantic "highlight" of sorts, but I find no way to do it with semantic highlights.  Next guess would be it is some sort of folding highlight, but that doesn't seem to be the case either.  Can anyone reveal how it is done?

**Comment by gjsjohnmurray** [comment, marked as answer]
Maybe TextEditor object's set Decorations method, using a whole-line decoration that specifies an opacity.

**Reply by dfgordon** [reply]
Thanks I'll try it!

**Comment by dfgordon** [comment]
Another way is to push a diagnostic with DiagnosticSeverity set to Hint, and with one DiagnosticTag set to Unnecessary.  The result of this is dimming without any squiggles, but with the hover text.
